// @aws-cdk/integ-tests-alpha/lib/assertions/providers/lambda-handler/http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpHandler = void 0;
/* eslint-disable no-console */
// eslint-disable-next-line import/no-extraneous-dependencies
const node_fetch_1 = require("node-fetch");
// TODO: can use builtin fetch on node18
const base_1 = require("./base");
class HttpHandler extends base_1.CustomResourceHandler {
    async processEvent(request) {
        console.log('request', request);
        const response = await (0, node_fetch_1.default)(request.parameters.url, request.parameters.fetchOptions);
        const result = {
            ok: response.ok,
            status: response.status,
            statusText: response.statusText,
            headers: response.headers.raw(),
        };
        result.body = await response.text();
        try {
            result.body = JSON.parse(result.body);
        }
        catch (e) {
            // Okay
        }
        return {
            apiCallResponse: result,
        };
    }
}
exports.HttpHandler = HttpHandler;
