// @aws-cdk/integ-tests-alpha/lib/assertions/providers/lambda-handler/types.js
"use strict";
// This file contains the input and output types for the providers.
// Kept in a separate file for sharing between the handler and the provider constructs.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Status = exports.AssertionType = exports.HTTP_RESOURCE_TYPE = exports.SDK_RESOURCE_TYPE_PREFIX = exports.ASSERT_RESOURCE_TYPE = void 0;
exports.ASSERT_RESOURCE_TYPE = 'Custom::DeployAssert@AssertEquals';
exports.SDK_RESOURCE_TYPE_PREFIX = 'Custom::DeployAssert@SdkCall';
exports.HTTP_RESOURCE_TYPE = 'Custom::DeployAssert@HttpCall';
/**
 * The type of assertion to perform
 */
var AssertionType;
(function (AssertionType) {
    /**
     * Assert that two values are equal
     */
    AssertionType["EQUALS"] = "equals";
    /**
     * The keys and their values must be present in the target but the target
     * can be a superset.
     */
    AssertionType["OBJECT_LIKE"] = "objectLike";
    /**
     * Matches the specified pattern with the array
     * The set of elements must be in the same order as would be found
     */
    AssertionType["ARRAY_WITH"] = "arrayWith";
})(AssertionType || (exports.AssertionType = AssertionType = {}));
/**
 * The status of the assertion
 */
var Status;
(function (Status) {
    /**
     * The assertion passed
     */
    Status["PASS"] = "pass";
    /**
     * The assertion failed
     */
    Status["FAIL"] = "fail";
})(Status || (exports.Status = Status = {}));
