// @aws-solutions-constructs/resources/test/template-writer.test.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const aws_cdk_lib_1 = require("aws-cdk-lib");
const assertions_1 = require("aws-cdk-lib/assertions");
const aws_s3_assets_1 = require("aws-cdk-lib/aws-s3-assets");
const path = require("path");
const template_writer_1 = require("../lib/template-writer");
test('TemplateWriter sets properties correctly', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const templateAsset = new aws_s3_assets_1.Asset(stack, 'TemplateAsset', {
        path: path.join(__dirname, 'template/sample-template')
    });
    const templateValues = new Array({
        id: 'placeholder',
        value: 'resolved_value'
    });
    template_writer_1.createTemplateWriterCustomResource(stack, 'Test', {
        templateBucket: templateAsset.bucket,
        templateKey: templateAsset.s3ObjectKey,
        templateValues
    });
    const cfnTemplate = assertions_1.Template.fromStack(stack);
    cfnTemplate.hasResourceProperties('Custom::TemplateWriter', {
        TemplateValues: JSON.stringify({ templateValues }),
        TemplateInputKey: templateAsset.s3ObjectKey,
        TemplateInputBucket: {
            "Fn::Sub": "cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}"
        },
        TemplateOutputBucket: {
            "Fn::Sub": "cdk-hnb659fds-assets-${AWS::AccountId}-${AWS::Region}"
        },
    });
});
test('TemplateWriter accepts custom lambda function timeout and memory size', () => {
    const stack = new aws_cdk_lib_1.Stack();
    const templateAsset = new aws_s3_assets_1.Asset(stack, 'TemplateAsset', {
        path: path.join(__dirname, 'template/sample-template')
    });
    const templateValues = new Array({
        id: 'placeholder',
        value: 'resolved_value'
    });
    template_writer_1.createTemplateWriterCustomResource(stack, 'Test', {
        templateBucket: templateAsset.bucket,
        templateKey: templateAsset.s3ObjectKey,
        templateValues,
        timeout: aws_cdk_lib_1.Duration.minutes(7),
        memorySize: 4096
    });
    const cfnTemplate = assertions_1.Template.fromStack(stack);
    cfnTemplate.hasResourceProperties('AWS::Lambda::Function', {
        Timeout: 420,
        MemorySize: 4096
    });
});
